components/batch-system/src/batch.rs

```diff
@@ -251,50 +242,40 @@ enum ReschedulePolicy {
 }
 
 impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
-    fn fetch_batch(&mut self, batch: &mut Batch<N, C>, max_size: usize) {
-        let curr_batch_len = batch.len();
-        if batch.control.is_some() || curr_batch_len >= max_size {
-            // Do nothing if there's a pending control fsm or the batch is already full.
-            return;
+    fn fetch_fsm(&mut self, batch: &mut Batch<N, C>) -> bool {
+        if batch.control.is_some() {
+            return true;
         }
 
-        let mut pushed = if curr_batch_len == 0 {
-            match self.fsm_receiver.try_recv().or_else(|_| {
-                self.handler.pause();
-                // Block if the batch is empty.
-                self.fsm_receiver.recv()
-            }) {
-                Ok(fsm) => batch.push(fsm),
-                Err(_) => return,
-            }
-        } else {
-            true
-        };
-
-        while pushed {
-            if batch.len() < max_size {
-                let fsm = match self.fsm_receiver.try_recv() {
-                    Ok(fsm) => fsm,
-                    Err(TryRecvError::Empty) => return,
-                    Err(TryRecvError::Disconnected) => unreachable!(),
-                };
-                pushed = batch.push(fsm);
-            } else {
-                return;
+        if let Ok(fsm) = self.fsm_receiver.try_recv() {
+            return batch.push(fsm);
+        }
+
+        if batch.is_empty() {
+            self.handler.pause();
+            if let Ok(fsm) = self.fsm_receiver.recv() {
+                return batch.push(fsm);
```

Comment:
As far as I know, the `TryRecvError::Disconnected` result returned by `try_recv` and the `RecvError` result returned by `recv` shouldn't happen in there, so we need to panic or assert to make logic more clearer